gossip.asciidoc

```diff
@@ -0,0 +1,655 @@
+;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43610
+;; flags: qr rd ra; QUERY: 1, ANSWER: 25, AUTHORITY: 0, ADDITIONAL: 1
+
+;; OPT PSEUDOSECTION:
+; EDNS: version: 0, flags:; udp: 512
+;; QUESTION SECTION:
+;test.nodes.lightning.directory.	IN	SRV
+
+;; ANSWER SECTION:
+test.nodes.lightning.directory.	59 IN	SRV	10 10 9735 ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7.test.nodes.lightning.directory.
+test.nodes.lightning.directory.	59 IN	SRV	10 10 15735 ln1qtgsl3efj8verd4z27k44xu0a59kncvsarxatahm334exgnuvwhnz8dkhx8.test.nodes.lightning.directory.
+
+<SNIP>
+
+;; Query time: 89 msec
+;; SERVER: 8.8.8.8#53(8.8.8.8)
+;; WHEN: Thu Dec 31 16:41:07 PST 2020
+```
+
+We've truncated the response for brevity. In our truncated responses, we can
+see two responses. Starting from the right-most column, we have a candidate
+"virtual" domain name for a target node, then to the left we have the _port_
+that this node can be reached using. The first response uses the standard port
+for LN: `9735`. The second response uses a custom port which is permitted by
+the protocol.
+
+Next, we'll attempt to obtain the other piece of information we need to connect
+to a node: it's IP address. Before we can query for this however, we'll fist
+_decode_ the returned sub-domain for this virtual host name returned by the
+server. To do that, we'll first encoded public key:
+```
+ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7
+```
+
+Using `bech32`, we can decode this public key to obtain the following valid
+`secp256k1` public key:
+```
+026c64f5a7f24c6f7f0e1d6ec877f23b2f672fb48967c2545f227d70636395eaf3
+```
+
+Now that we have the raw public key, we'll now ask the DNS server to _resolve_
+the virtual host given so we can obtain the IP information for the node:
+```
+$ dig ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7.test.nodes.lightning.directory A
+
+; <<>> DiG 9.10.6 <<>> ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7.test.nodes.lightning.directory A
+;; global options: +cmd
+;; Got answer:
+;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41934
+;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
+
+;; OPT PSEUDOSECTION:
+; EDNS: version: 0, flags:; udp: 4096
+;; QUESTION SECTION:
+;ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7.test.nodes.lightning.directory. IN A
+
+;; ANSWER SECTION:
+ln1qfkxfad87fxx7lcwr4hvsalj8vhkwta539nuy4zlyf7hqcmrjh40xx5frs7.test.nodes.lightning.directory. 60 IN A X.X.X.X
+
+;; Query time: 83 msec
+;; SERVER: 2600:1700:6971:6dd0::1#53(2600:1700:6971:6dd0::1)
+;; WHEN: Thu Dec 31 16:59:22 PST 2020
+;; MSG SIZE  rcvd: 138
+```
+
+In the above command, we've queried the server so we can obtain an `IPv4`
+address for our target node. Now that we have both the raw public key _and_ IP
+address, we can connect to the node using the `brontide` transport protocol at:
+`026c64f5a7f24c6f7f0e1d6ec877f23b2f672fb48967c2545f227d70636395eaf3@X.X.X.X`.
```

Comment:
It would be great if we can expand a little bit more here. How is this public key being used? I am assuming the public key is used for secure communcation, am I right? Can the "client" trust it has been given the correct public key going from the DNS? In a HTTP/browser world we never trust the dns, but always trust the CA's. I'm thinking there might be mitm-attacks here if the public key is used as a basis for KDF/encryption. What do you think @Roasbeef ?

Reply:
@aantonop and @Roasbeef, as I understand it this public key is used with ECDH and encryption handshakes https://github.com/lightningnetwork/lightning-rfc/blob/b201efe0546120c14bf154ce5f4e18da7243fe7a/08-transport.md#authenticated-key-agreement-handshake. Can the user fully trust that the unencrypted dns query is giving the correct public key and destination? How can we gain trust in this case?